### PR TITLE
Anesthesia Real

### PR DESCRIPTION
**Pull request by IProduceWidgets**
## About the PR

Implanters now require the target is asleep to remove an implant.
Also some minor bugfix for nitrous and freezon status effects. They now correctly persist on continued use.

## Why / Balance

Require extra effort for implant checking.

## Technical details

Adds a do after option to require the target is asleep.

## Media

anesthesia

- [X] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

## Breaking changes

**Changelog**

:cl:
- tweak: Implant removal now requires anesthesia for all species but slimes!

### CODE REVIEW

**Comment by Nairodian** [comment]
If it's going to require sleeping, I'd decrease the length of the do-after just to be less of a pain in the ass.

**Comment by notquitehadouken** [comment]
surgery update????????

**Comment by Holinka4ever** [comment]
Just interesting, how it will work on humanoids that are dead, cause there are contorversal things like:

1)If removing implant from a dead person is not possible without being a patient alive+sleep state, it will be non-logical as hell.

2)if removing implant from a dead person is possible, it will just encourage sec to kill a target, that it will give a frustration to antag/non-antag patient.

**Comment by Hmeister-real** [comment]
>2)if removing implant from a dead person is possible, it will just encourage sec to kill a target, that it will give a frustration to antag/non-antag patient.

I doubt sec will start killing people who are AoS on sight just to check for implants, if they do that isn't a powergaming issue it's an admin issue. Unless they are a confirmed syndie and it's lrp, then I can see that happening.

**Comment by Anzarot121** [comment]
a person can simply turn it off using the sign on the right.

**Comment by IProduceWidgets** [comment]
> 1)If removing implant from a dead person is not possible without being a patient alive+sleep state, it will be non-logical as hell.

Right now it requires the sleeping component, which you lose when you die. I'm not going to make it work on dead people, as someone else is currently working on consciousness states and it would overlap a bit too much.

> a person can simply turn it off using the sign on the right.

I'm not 100% what you mean, but if you mean players can wake themselves up, this PR also stops that from working with nitrous oxide. (It also is prevented by chloral and nocturine, but in a small enough time span that its probably not useful here unless you have a second person administering it throughout the removal -- maybe once IV bags are added that would work.)

You do however have the option to go to sleep to have your implant removed, if for some reason you wanted that. Nitrous just allows for forced sleeping.

**Comment by IProduceWidgets** [comment]
I decreased the draw time to 30s from 60s since the time difference is made up and more by the other doafters needed from stripping / internals etc.

**Comment by Emisse** [comment]
Id like Kerons thoughts on this before proceeding

**Comment by Geekyhobo** [comment]
I think this can be meta gamed by just clicking on someone with an implanter when they are not asleep, and getting the message that they can't removed the implant i think its a cool feature but it might be more logical to wait til surgery is a thing before we merge this. I think it would work well and be pretty cool if the only way to remove or check for an implanter is when they are undergoing surgery, also as i was writing this i realized you should just have it say the patent must be asleep to perform this action. just don't say the work implanter because it might make people have a bias to check for an implant because the wording leads them to believe the implant regardless of the person having it or not.

**Comment by IProduceWidgets** [comment]
> I think this can be meta gamed by just clicking on someone with an implanter when they are not asleep, and getting the message that they can't removed the implant i think its a cool feature but it might be more logical to wait til surgery is a thing before we merge this.

Nah, it always fails if they are asleep, and otherwise follows the current logic of nothing happening if they do not have an implant once the doafter completes.

I'll change the popup text. I was unsure of how to word it, and was more focused on avoiding the identity system.

**Comment by IProduceWidgets** [comment]
Test fail is erroneous

**Comment by Anzarot121** [comment]
Found another problem. Slime people don't sleep from nitric oxide.

**Comment by Aisu9** [comment]
> Found another problem. Slime people don't sleep from nitric oxide.

another use for hydrochloric hydrate

**Comment by keronshb** [comment]
> Id like Kerons thoughts on this before proceeding

I don't know if I like the change or not.

The extremely long timer for removal was originally there as a placeholder until surgery was in.

If this helps as an extra step I don't see the harm but I don't exactly like it either.

**Comment by Simyon264** [comment]
I like this change. It adds a bit of RP possibilities.

**Comment by Emisse** [comment]
goncflicts

**Comment by IProduceWidgets** [comment]
I fixed the conflict, it was just someone fixing nitrous in another PR the same way I did here.

I have no good answer to the slime issue though. I also don't know how surgery will handle that either though, so...

**Comment by K-Dynamic** [comment]
> I fixed the conflict, it was just someone fixing nitrous in another PR the same way I did here.
> 
> I have no good answer to the slime issue though. I also don't know how surgery will handle that either though, so...

Guess the surgeon will just have to operate on slime patients while they're awake (i.e. make this pr's implant removers work on awake slimes)

That or:
- use chloral hydrate
- wait for nocturine recipe to be added as part of chem rework

**Comment by deltanedas** [comment]
nocturine recipe currently requires a toxin kit or an emag, both are ideally out of reach of rule abiding goodsec

**Comment by IProduceWidgets** [comment]
I made it work on slimes that are awake, the solution isn't elegant though because its not generalized for the doafter and instead just special cases the slime species in the implanter system. 

Its the best I can come up with though.

**Review by metalgearsloth — CHANGES_REQUESTED** [review]
Use spaces not tabs please, mixing it is hell and will immediately get diffed by anyone else who touches the file.

image

**Comment by IProduceWidgets** [comment]
I refuse to believe that removing tabs caused that test fail.

File: Content.Server/Implants/ImplanterSystem.cs [hunk 1]
```diff
@@ -103,17 +106,36 @@ public void TryImplant(ImplanterComponent component, EntityUid user, EntityUid t
...
+                BreakOnTargetMove = true,
+                BreakOnDamage = true,
+			    BreakOnTargetWake = false,
+                NeedHand = true,
+            };
+            
+            if (_doAfter.TryStartDoAfter(args))
+            _popup.PopupEntity(Loc.GetString("injector-component-injecting-user"), target, user);
+        }
     }
```
Comment (metalgearsloth) [inline Content.Server/Implants/ImplanterSystem.cs:139]: 1. Don't mix tabs and spaces.
2. Don't hardcode this for species, use a component if yo ureally want it.
3. Please don't inline your ifs like this it makes the formatting a mess.

File: Content.Shared/DoAfter/DoAfterArgs.cs [hunk 2]
```diff
@@ -127,6 +127,12 @@ public sealed partial class DoAfterArgs
     /// </summary>
     [DataField("breakOnDamage")]
     public bool BreakOnDamage;
+    
+    /// <summary>
+    ///     Whether a change to the target sleep state will cancel the DoAfter.
+    /// </summary>
+    [DataField("breakOnTargetWake")]
+    public bool BreakOnTargetWake;
```
Comment (metalgearsloth) [inline Content.Shared/DoAfter/DoAfterArgs.cs:136]: Datafield strings are obsolete.

File: Resources/Prototypes/Reagents/gases.yml [hunk 3]
```diff
@@ -254,7 +254,8 @@
           shouldHave: false
         key: ForcedSleep
         component: ForcedSleeping
-        time: 3
+        time: 250 # time in miliseconds (due to how respiration works this needs to be over 2 seconds to prevent losing the effect.)
+        refresh: true
```
Comment (metalgearsloth) [inline Resources/Prototypes/Reagents/gases.yml:258]: You can't just say it's milliseconds and have it work, this is deserialized as seconds.

**Review by metalgearsloth — CHANGES_REQUESTED** [review]

**Comment by metalgearsloth** [comment]
Re-open if you come back to it.

**Pull request closed without merging**